### PR TITLE
[CBRD-23026] Heap append pages

### PR DESCRIPTION
**Pull request by lucianlazar2211**
http://jira.cubrid.org/browse/CBRD-23026

This function is able to append a set of pages to a given heap by creating a `HEAP_CHAIN` using those pages. Then we append the heap chain to the heap by fixing the `heap_header_page`, followed by fixing the `last_page` of the heap. Then we fix the first page of the heap chain and we create the new links.

### CODE REVIEW

File: src/storage/heap_file.c [hunk 1]
```diff
@@ -24743,5 +24748,265 @@ heap_nonheader_page_capacity ()
...
+  if (chain == NULL)
+    {
+      // This should never happen
+      assert (false);
+      error_code = ER_FAILED;
+      return error_code;
+    }
+
+  // Make sure we fixed the page.
+  assert (pgbuf_is_page_fixed_by_thread (thread_p, vpid));
```
Comment (razvanradulescu) [inline src/storage/heap_file.c]: You already have the same  check some lines above; I don't think is necessary again here :)

File: src/storage/heap_file.c [hunk 2]
```diff
@@ -24743,5 +24748,265 @@ heap_nonheader_page_capacity ()
...
+   if (heap_header_watcher.pgptr)
+    {
+      pgbuf_ordered_unfix_and_init (thread_p, heap_header_watcher.pgptr, &heap_header_watcher);
+    }
+
+   return error_code;
+}
+
+static int
+heap_get_page_with_watcher (THREAD_ENTRY * thread_p, VPID *page_vpid, PGBUF_WATCHER * pg_watcher)
```
Comment (razvanradulescu) [inline src/storage/heap_file.c]: const for page_vpid

File: src/storage/heap_file.c [hunk 3]
```diff
@@ -24743,5 +24748,265 @@ heap_nonheader_page_capacity ()
...
+    {
+      ASSERT_ERROR_AND_SET (error_code);
+      return error_code;
+    }
+
+   return error_code;
+}
+
+static int
+heap_add_chain_links (THREAD_ENTRY * thread_p, const HFID * hfid, VPID * vpid, VPID * next_link, VPID * prev_link,
```
Comment (razvanradulescu) [inline src/storage/heap_file.c]: const for vpid, next_link, prev_link

File: src/storage/heap_file.c [hunk 4]
```diff
@@ -24743,5 +24748,265 @@ heap_nonheader_page_capacity ()
...
+  log_append_undoredo_data (thread_p, RVHF_STATS, &addr, sizeof(HEAP_HDR_STATS), sizeof (HEAP_HDR_STATS), 
+                            &heap_hdr_prev, heap_hdr);
+
+  // Set the page as dirty.
+  pgbuf_set_dirty (thread_p, heap_header_watcher.pgptr, DONT_FREE);
+
+   // Now we can commit the sysop.
+  log_sysop_commit (thread_p);
+
+cleanup:
```
Comment (eseokoh) [inline src/storage/heap_file.c]: I'd like to add this to make it sure:
```c
if (error_code != NO_ERROR)
  {
    log_sysop_abort (thread_p);
  }
```

We may change with above commit as follows:
```c
if (error_code != NO_ERROR)
  {
    log_sysop_abort (thread_p);
  }
else
  {
    log_sysop_commit (thread_p);
  }
```

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/heap_file.c (2 incorrect indentation(s) found)
```
   885 				 const VPID * prev_link, PGBUF_WATCHER * page_watcher, bool keep_page_fixed,
   886 				 bool is_page_watcher_inited);
```
#### File: src/storage/heap_file.h (3 incorrect indentation(s) found)
```
   673 extern int heap_alloc_new_page (THREAD_ENTRY * thread_p, HFID * hfid, OID class_oid, PGBUF_WATCHER * home_hint_p,
   674 				VPID * new_page_vpid);
   676 extern int heap_nonheader_page_capacity ();
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/heap_file.c (2 incorrect indentation(s) found)
```
   885 				 const VPID * prev_link, PGBUF_WATCHER * page_watcher, bool keep_page_fixed,
   886 				 bool is_page_watcher_inited);
```
#### File: src/storage/heap_file.h (3 incorrect indentation(s) found)
```
   673 extern int heap_alloc_new_page (THREAD_ENTRY * thread_p, HFID * hfid, OID class_oid, PGBUF_WATCHER * home_hint_p,
   674 				VPID * new_page_vpid);
   676 extern int heap_nonheader_page_capacity ();
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/storage/heap_file.c [hunk 5]
```diff
@@ -24743,5 +24748,287 @@ heap_nonheader_page_capacity ()
...
+                                     &heap_last_page_watcher, true, true);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Now update the last page of the heap header.
+
+  heap_hdr->estimates.last_vpid = heap_pages_array[array_size - 1];
```
Comment (ribram) [inline src/storage/heap_file.c]: Could you add a separate heap function to update&log last_vpid/count in header page?

File: src/storage/heap_file.c [hunk 6]
```diff
@@ -24743,5 +24748,287 @@ heap_nonheader_page_capacity ()
...
+      pgbuf_ordered_unfix_and_init (thread_p, heap_last_page_watcher.pgptr, &heap_last_page_watcher);
+    }
+
+  if (heap_header_watcher.pgptr)
+    {
+      pgbuf_ordered_unfix_and_init (thread_p, heap_header_watcher.pgptr, &heap_header_watcher);
+    }
+
+  // Check if we have errors to abort the sysop.
+  if (error_code != NO_ERROR)
```
Comment (ribram) [inline src/storage/heap_file.c]: Sysop should be committed/aborted before unfixing header/last pages; if pages are unfixed first, other changes are possible although current changes are not yet persistent and can be undone at recovery.

File: src/storage/heap_file.c [hunk 7]
```diff
@@ -24743,5 +24748,287 @@ heap_nonheader_page_capacity ()
...
+      error_code = heap_get_page_with_watcher (thread_p, vpid, page_watcher);
+      if (error_code != NO_ERROR)
+        {
+          ASSERT_ERROR ();          
+          return error_code;
+        }
+    }
+
+  // Make sure we fixed the page.
+  assert (pgbuf_is_page_fixed_by_thread (thread_p, (VPID *) vpid));
```
Comment (ribram) [inline src/storage/heap_file.c]: Cast won't be necessary if pgbuf_is_page_fixed_by_thread signature is changed to accept `const VPID *`.

File: src/transaction/locator_sr.c [hunk 8]
```diff
@@ -13711,17 +13711,14 @@ locator_multi_insert_force (THREAD_ENTRY * thread_p, HFID * hfid, OID * class_oi
...
 {
   int error_code = NO_ERROR;
   size_t accumulated_records_size = 0;
   size_t heap_max_page_size;
   OID dummy_oid;
-  // *INDENT-OFF*
   std::vector<RECDES> recdes_array;
   std::vector<VPID> heap_pages_array;
-  // *INDENT-ON*
+
```
Comment (ribram) [inline src/transaction/locator_sr.c]: remove one empty line

File: src/storage/heap_file.c (hunk 7 above)
Comment (ribram) [reply src/storage/heap_file.c]:
```suggestion
  assert (pgbuf_is_page_fixed_by_thread (thread_p, vpid));
```

File: src/transaction/locator_sr.c [hunk 9]
```diff
@@ -13797,24 +13794,30 @@ locator_multi_insert_force (THREAD_ENTRY * thread_p, HFID * hfid, OID * class_oi
...
+        {
 	  ASSERT_ERROR ();
 	  return error_code;
-	}
+        }
     }
 
   // Now form a heap chain with the pages and add the chain to the current heap.
-  // TODO: this!
+  error_code = heap_append_pages_to_heap (thread_p, hfid, (const OID &) class_oid, heap_pages_array);
```
Comment (ribram) [inline src/transaction/locator_sr.c]:
```suggestion
  error_code = heap_append_pages_to_heap (thread_p, hfid, *class_oid, heap_pages_array);
```

**Pull request merged**